cmd/kubeadm/app/cmd/phases/upgrade/apply/uploadconfig.go

```diff
@@ -107,9 +108,11 @@ func runUploadKubeletConfig(c workflow.RunData) error {
 		return errors.Wrap(err, "error creating kubelet configuration ConfigMap")
 	}
 
-	klog.V(1).Infoln("[upgrade/upload-config] Preserving the CRISocket information for this control-plane node")
-	if err := patchnodephase.AnnotateCRISocket(client, cfg.NodeRegistration.Name, cfg.NodeRegistration.CRISocket); err != nil {
-		return errors.Wrap(err, "error writing Crisocket information for the control-plane node")
+	if !features.Enabled(cfg.ClusterConfiguration.FeatureGates, features.NodeLocalCRISocket) {
+		klog.V(1).Infoln("[upgrade/upload-config] Preserving the CRISocket information for this control-plane node")
+		if err := patchnodephase.AnnotateCRISocket(client, cfg.NodeRegistration.Name, cfg.NodeRegistration.CRISocket); err != nil {
+			return errors.Wrap(err, "error writing Crisocket information for the control-plane node")
+		}
```

Comment:
what about:
kubeadm upgrade node
kubeadm reset

they also need to handle the annotation vs file, correct?

Reply:
Perhaps we should only stop setting this annotation at GA, otherwise, we need to continue setting it during kubeadm init, join, and upgrade, right?

If we don't do this, we won't be able to ensure compatibility for users when they enable and disable the feature gate.

Reply:
adding a follow up comment

Reply:
Implementation of kubeadm reset has been added.

I think `kubeadm upgrade node` should be the same as `kubeadm upgrade apply`, and you don't need to care about annotations. 
* If the kubeadm-flags.env file contains the `--container-runtime-endpoint` flag, we write it to `instance-config.yaml`. 
* If the `kubeadm-flags.env` file does not contain the `--container-runtime-endpoint` flag, `instance-config.yaml` should already exist. 
* If the `kubeadm-flags.env` file does not contain the `--container-runtime-endpoint` flag and `instance-config.yaml` does not exist, we have entered the wrong area and we use `DefaultCRISocket`.